### PR TITLE
offset_from: do not require pointers to be inbounds

### PR DESCRIPTION
**Pull request by RalfJung**
In `offset_from`, we currently require both pointers to be *in-bounds* of the same allocation.

There's not really any good motivation for this requirement: all codegen backends implement this intrinsic by casting to integers and doing the subtraction there, and const-eval (the original motivation for having this function in the first place) is fine as long as both pointers are *derived from* the same allocation, but doesn't care if they are in-bounds.

This restriction also causes problems, as noted in https://github.com/rust-lang/rust/issues/92512. We allow CTFE to move pointers out-of-bounds with `wrapping_offset`, but it is currently impossible to determine how far they have been moved, as `offset_from` is the only "pointer subtraction" operation available in `const` and it has this restriction on things being in-bounds.

Hence I propose we remove this restriction. @rust-lang/opsem what do you think? (For FCP we'll probably want to add T-lang since this is stable.)

The one interesting question this open is how to interpret the requirement about not wrapping the edge of the address space in CTFE, where we don't know where that edge is. Currently we basically pretend that the edge is at the beginning of the allocation, so a pointer to 4 bytes before that beginning is considered to "wrap" when used with a pointer to 4 bytes after that beginning.

The Miri implementation also needs some work for wildcard pointers (as returned by int2ptr casts). But I figured I'd see if we even want this change before implementing all that.

### CODE REVIEW

**Comment by the8472** [comment]
How does this affect `sub_ptr`? It has its own intrinsic but its documentation points to the `offset_from` requirements.

**Comment by JakobDegen** [comment]
@RalfJung it's not quite clear to me how I'm expected to be able to use this feature correctly in a const context. If I offset a pointer to outside its allocation, is it ever possible for me to know that such an operation won't overflow the address space? That does happen to be the case right now because all const allocations are at the beginning of the address space, but I don't think that's stable.

Unrelated: Does having the "same allocations" restriction as a safety requirement make sense? It seems like we could 1. drop this requirement outside of const contexts, and 2. promise to throw an error and halt compilation if this requirement is violated in a const context. That seems like a much more user friendly version of things than UB

**Comment by RalfJung** [comment]
Yeah I realized what I really want is a wrapping_offset_from that still requires "based on same alloc" but makes no overflow restrictions.

Regarding the second point, as of now we that would be an option though we don't have precedent for such an operation. However, the "based on same alloc" is actually useful for optimizations; it means this operation does not "escape" the provenance (in an alias analysis sense, not an ptr2int semantics sense).

**Comment by JakobDegen** [comment]
I realize I didn't say it before, but I should clarify: Even with my two above points (which I do think should be considered) this PR doesn't make things worse and I'm in support regardless.

> Regarding the second point, as of now we that would be an option though we don't have precedent for such an operation. However, the "based on same alloc" is actually useful for optimizations; it means this operation does not "escape" the provenance (in an alias analysis sense, not an ptr2int semantics sense).

I see the point in theory, but I can't actually think of an example where it's useful. Might be worth taking to a separate Zulip thread though?

File: library/core/src/ptr/const_ptr.rs [hunk 1]
```diff
@@ -644,6 +642,12 @@ impl<T: ?Sized> *const T {
     /// (Note that [`offset`] and [`add`] also have a similar limitation and hence cannot be used on
     /// such large allocations either.)
     ///
+    /// The requirement for pointers to be derived from the same allocated object is primarily
+    /// equired for `const`-compatibility: at compile-time, pointers into *different* allocated
```
Comment (yotamofek) [inline library/core/src/ptr/const_ptr.rs:646]:
```suggestion
    /// required for `const`-compatibility: at compile-time, pointers into *different* allocated
```
😄

**Comment by RalfJung** [comment]
> I realize I didn't say it before, but I should clarify: Even with my two above points (which I do think should be considered) this PR doesn't make things worse and I'm in support regardless.

I tend to disagree. If we add `wrapping_offset_from` as described above, we have a reasonable symmetry:
- if `ptr2` was obtained from `ptr1` via `offset`, then `offset_from` is legal, and conversely if `offset_from` is legal then obtaining `ptr2` from `ptr` via `offset` would be possible
- if `ptr2` was obtained from `ptr1` via `wrapping_offset`, then `wrapping_offset_from` is legal, and conversely if `wrapping_offset_from` is legal then obtaining `ptr2` from `ptr` via `wrapping_offset` would be possible

To maintain that symmetry we should drop the `inbounds` requirement from *both* `offset` and `offset_from`, not just one of them.

**Comment by RalfJung** [comment]
This PR is superseded by https://github.com/rust-lang/rust/pull/112837.

**Pull request closed without merging**